### PR TITLE
[NFT Metadata Crawler] GCS Write + Postgres Commit

### PR DESCRIPTION
**Pull request by just-in-chang**
### Description

Functionality for writing to GCS and handle errors for Postgres commits

Fixes EI-261

### CODE REVIEW

File: ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/parser_entry.rs [hunk 1]
```diff
@@ -168,21 +174,67 @@ impl ParserEntry {
     }
 
     /// Calls and handles error for writing JSON to GCS
-    async fn handle_write_json_to_gcs(&mut self, _json: Option<Value>) -> Option<String> {
-        todo!();
+    async fn handle_write_json_to_gcs(&mut self, json: Option<Value>) -> Option<String> {
```
Comment (bowenyang007) [inline ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/parser_entry.rs:177]: If when json is null, you'll log an error, then just log the error before getting into this function.

File: ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/parser_entry.rs [hunk 2]
```diff
@@ -168,21 +174,67 @@ impl ParserEntry {
...
+            match write_json_to_gcs(
+                self.token.clone(),
+                self.bucket.clone(),
+                self.entry.token_data_id.clone(),
+                json,
+            )
+            .await
+            {
+                Ok(filename) => {
+                    self.log_info("Successfully wrote JSON to GCS");
```
Comment (bowenyang007) [inline ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/parser_entry.rs:188]: btw why are you implementing your own log wrapper? Isn't tracing::info! enough? 

Another thing is that your logs are basically useless as is b/c there is no metadata. We need things like the token_data_id, bucket, etc. Please use structured logs as well, e.g. 

tracing::info!(token_data_id = self.entry.token_data_id, "blabla");

File: ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/parser_entry.rs [hunk 3]
```diff
@@ -168,21 +174,67 @@ impl ParserEntry {
...
+            self.log_error("No JSON to write to GCS");
+            None
+        }
     }
 
     /// Calls and handles error for writing image to GCS
     async fn handle_write_image_to_gcs(
         &mut self,
-        _image: Option<(Vec<u8>, ImageFormat)>,
+        image: Option<(Vec<u8>, ImageFormat)>,
```
Comment (bowenyang007) [inline ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/parser_entry.rs:205]: same here, let's not pass in optional if the function doesn't actualyl ike it.

File: ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/parser_entry.rs [hunk 4]
```diff
@@ -168,21 +174,67 @@ impl ParserEntry {
...
     }
 
     /// Calls and handles error for upserting to Postgres
     async fn commit_to_postgres(&mut self) {
-        todo!();
+        match upsert_uris(&mut self.conn, self.model.clone()) {
+            Ok(_) => self.log_info("Successfully committed to Postgres"),
+            Err(e) => self.log_error(&format!("Failed to commit to Postgres: {}", e)),
+        };
     }
```
Comment (bowenyang007) [inline ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/parser_entry.rs:238]: You'll find that this error logging won't help you that much without the query itself. Hence I was telling you to use something like this 
https://github.com/bowenyang007/aptos-core/blob/main/crates/indexer/src/database.rs#L64-L89

File: ecosystem/nft-metadata-crawler/nft-metadata-crawler-utils/src/gcs.rs [hunk 5]
```diff
@@ -1,23 +1,92 @@
...
+    id: String,
+    json: Value,
 ) -> anyhow::Result<String> {
-    todo!();
+    let client = Client::new();
+    let filename = format!("json_{}.json", id);
+    let url = format!(
+        "https://storage.googleapis.com/upload/storage/v1/b/{}/o?uploadType=media&name={}",
+        bucket, filename
+    );
```
Comment (bowenyang007) [inline ecosystem/nft-metadata-crawler/nft-metadata-crawler-utils/src/gcs.rs:22]: this part is the same for below as well, copy these into a different function.

File: ecosystem/nft-metadata-crawler/nft-metadata-crawler-utils/src/gcs.rs [hunk 6]
```diff
@@ -1,23 +1,92 @@
...
+    let client = Client::new();
+    let mut headers = HeaderMap::new();
+
+    let extension = match img_format {
+        ImageFormat::Gif | ImageFormat::Avif => img_format
+            .extensions_str()
+            .last()
+            .unwrap_or(&"gif")
+            .to_string(),
+        _ => "jpeg".to_string(),
```
Comment (jjleng) [inline ecosystem/nft-metadata-crawler/nft-metadata-crawler-utils/src/gcs.rs:59]: why no png here?

Comment (just-in-chang) [reply ecosystem/nft-metadata-crawler/nft-metadata-crawler-utils/src/gcs.rs:59]: Converting all PNGs to JPEGs so they are smaller 🙂

File: ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/parser_entry.rs [hunk 7]
```diff
@@ -183,7 +188,16 @@ impl ParserEntry {
                 last_transaction_version = self.entry.last_transaction_version,
                 "Committing image optimization to Postgres"
             );
-            self.commit_to_postgres().await;
+            match upsert_uris(&mut self.conn, self.model.clone()) {
+                Ok(_) => info!(
+                    last_transaction_version = self.entry.last_transaction_version,
```
Comment (bowenyang007) [inline ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/parser_entry.rs:193]: let's log the result.

File: ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/utils/db.rs [hunk 8]
```diff
@@ -27,16 +28,9 @@ pub fn upsert_uris(
...
-        },
-        Err(e) => {
-            error!("Failed to upsert URIs: {}", e.to_string());
-            Err(e.into())
-        },
-    }
+        ));
+
+    let debug_query = diesel::debug_query::<diesel::pg::Pg, _>(&query).to_string();
+    info!("Executing Query: {}", debug_query);
```
Comment (bowenyang007) [inline ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/utils/db.rs:34]: debug

**Review by bowenyang007 — APPROVED** [review]

**Review by jjleng — APPROVED** [review]

**Pull request merged**